src/extension.ts

```diff
@@ -64,37 +61,46 @@ function activateMdExt(context: ExtensionContext) {
         wordPattern: /(-?\d*\.\d\w*)|([^\!\@\#\%\^\&\(\)\-\=\+\[\{\]\}\\\|\;\:\'\"\,\.\<\>\/\?\s\，\。\《\》\？\；\：\‘\“\’\”\（\）\【\】\、]+)/g
     });
 
-    newVersionMessage(context.extensionPath);
+    showWelcome(context);
 }
 
-function newVersionMessage(extensionPath: string) {
-    let data: string, currentVersion: string;
+/**
+ * Shows a welcome message on first time startup.
+ */
+async function showWelcome(context: ExtensionContext): Promise<void> {
+    // The directory for an extension is recreated every time VS Code installs it.
+    // Thus, we only need to read and write an empty flag file there.
+    // If the file exists, then it's not the first time, and we don't need to do anything.
+    const flagFileUri = Uri.joinPath(context.extensionUri, "WELCOMED");
     try {
-        data = fs.readFileSync(`${extensionPath}${path.sep}package.json`).toString();
-        currentVersion = JSON.parse(data).version;
-        if (
-            fs.existsSync(`${extensionPath}${path.sep}VERSION`)
-            && fs.readFileSync(`${extensionPath}${path.sep}VERSION`).toString() === currentVersion
-        ) {
-            return;
-        }
-        fs.writeFileSync(`${extensionPath}${path.sep}VERSION`, currentVersion);
-    } catch (error) {
-        console.log(error);
+        await workspace.fs.stat(flagFileUri);
         return;
+    } catch {
+        workspace.fs.writeFile(flagFileUri, new Uint8Array()).then(() => { }, () => { });
     }
-    const featureMsg = getNewFeatureMsg(currentVersion);
-    if (featureMsg === undefined) return;
-    const message1 = localize("showMe");
-    const message2 = localize("dismiss");
-    window.showInformationMessage(featureMsg, message1, message2).then(option => {
-        switch (option) {
-            case message1:
-                showChangelog();
-            case message2:
-                break;
+
+    // The existence of welcome materials depends on build options we set during pre-publish.
+    // If any condition is not met, then we don't need to do anything.
+    try {
+        const welcomeMessageFileUri = Uri.joinPath(context.extensionUri, "welcome.txt");
+        const msgWelcome = Buffer.from(await workspace.fs.readFile(welcomeMessageFileUri)).toString("utf8");
+        if (/^\s*$/.test(msgWelcome) || /\p{C}/u.test(msgWelcome)) {
+            return;
         }
-    });
+
+        const changelogFileUri = Uri.joinPath(context.extensionUri, "changes.md");
+        await workspace.fs.stat(changelogFileUri);
+
+        const btnOpenLocal = localize("ui.welcome.buttonOpenLocal");
+
+        window.showInformationMessage(msgWelcome, btnOpenLocal).then(selection => {
+            switch (selection) {
+                case btnOpenLocal:
+                    workspace.openTextDocument(changelogFileUri).then(window.showTextDocument);
+                    return;
```

Comment:
Let's still include a "Dismiss"/"Cancel" button here. I believe (for many people) it is very unpleasant to click on the small "x".

![image](https://user-images.githubusercontent.com/7588612/105376245-81ca9180-5c44-11eb-9804-b0d8e9a1c643.png)